terminus/geometry/circle.py

```diff
@@ -45,45 +45,66 @@ def __eq__(self, other):
     def __hash__(self):
         return hash((self.center, self.radius))
 
+    def radius(self):
+        return self.radius
+
+    def center(self):
+        return self.center
+
     def almost_equal_to(self, other, decimals=5):
         return self.center.almost_equal_to(other.center, decimals) and abs(self.radius - other.radius) < 1e-5
 
-    def intersection(self, other):
-        d = (other.center - self.center).norm()
-        r1 = self.radius
-        r2 = other.radius
-        if d == 0 and r1 == r2:
+    def intersection(self, other, buffer=1e-7, decimals=7):
+        p = self.center
+        q = other.center
+        A = self.radius
+        B = other.radius
+        """
+        To find the intersection, we solve the two equations system given by:
+        (Point(x, y) - p).norm_squared = A ** 2
+        (Point(x, y) - q).norm_squared = B ** 2
+        """
+        M = 2 * (q.x - p.x)
+        N = 2 * (q.y - p.y)
+        R = A ** 2 - B ** 2 + q.norm_squared() - p.norm_squared()
+        # M * x + N * y = R
+        if self.almost_equal_to(other, decimals):
             return [self]
-        if d == 0 and r1 != r2:
+        if p == q and A != B:
             return []
-        if (d > self.radius + other.radius) or (d < abs(self.radius - other.radius)):
+        if abs(N) < buffer:
+            x = R / M
+            if abs(x - p.x) < A and abs(x - q.x) < B:
```

Comment:
It is not about giving a longer name. What we are looking for is:

- Meaningful names (`A` is not meaningful, as you need to keep it on some place in your head that you assigned to it `self.radius` 20, 30 or 40 lines before. You are forcing the reader to use his/her brain to store some information instead of focusing on the algorithm itself. Things like `my_radius` are good because they are self-descriptive; you don't even need to know where or when it was assigned, you just know what it is).
- Avoiding computing the same thing many times. As an example, `A ** 2` is computed 4 times, when you can just compute it one time and store it for later use.